shadertoy-plugin.js

```diff
@@ -381,6 +400,121 @@
         }
     };
 
+    /**
+     * Create the UI controls to download a ZIP of the current shader
+     */
+    ToyPlugEditPage.prototype.downloadShaderAsZip = function downloadShaderAsZip() {
+        /**
+         * Prepare and download a ZIP of the current shader
+         */
+        function onDownloadAsZipClick() {
+
+            // Get names we'll need
+            var id = gShaderToy.mInfo.id;
+            if (id == '-1') {
+                id = 'default';
+            }
+
+            var friendly_name =
+                gShaderToy.mInfo.name ||
+                ('ShaderToy shader ' + id);
+
+            var username = gShaderToy.mInfo.username || 'unknown';
+
+            var filename =
+                sanitize_filename(username) + ' - ' +
+                sanitize_filename(friendly_name);
+
+            //Make readme.txt
+            var readme =
+                '"' + friendly_name + '" provided by ' + username + '\n';
+            readme += window.location.href + '\n';
+            readme += '-'.repeat(78) + '\n';
+
+            // Pull from the live tags and description fields, if provided,
+            // rather than just from mInfo.  This permits
+            // stuffing information (e.g., release party) into the
+            // readme without having to hand-edit, and without needing
+            // write access to the shader.
+
+            // Get the tags from the tag box, with a fallback to the
+            // ShaderToy info.
+            var tags = [];
+            if(shaderToyElements.shaderTags &&
+                                    shaderToyElements.shaderTags.value) {
+                tags = shaderToyElements.shaderTags.value.split(/\s*,\s*/);
+            }
+            if(tags.length == 0) tags = gShaderToy.mInfo.tags;
+
+            // Add the tags to the README
+            if(Array.isArray(tags) && tags.length != 0) {
+                readme += 'Tags: ' + tags.join(', ') + '\n';
+            }
+
+            // Get the description from the textarea, with a fallback
+            // to the ShaderToy info
+            var description;
+            if(shaderToyElements.shaderDescription &&
+                            shaderToyElements.shaderDescription.value) {
+                description = shaderToyElements.shaderDescription.value;
+            } else {
+                description = gShaderToy.mInfo.description;
+            }
+
+            // Add the description to the README
+            if(description) {
+                readme += '\n' + description + '\n';
+            }
+
+            // Add a usage reminder
+            readme += '\n' + '-'.repeat(78) + '\n';
```

Comment:
please use template for readme contents,

Reply:
Which type of template?

 - ES6 template literals (Chrome 41+, according to MDN)
 - a string literal (or XHR from a separate file in the extension) I can regex-replace, 
 - a template engine (I doubt it)
 - something else

Reply:
i meant smth like this:

```
README_CONTENT_TEMPLATE = '---------------- {shaderName} --------\n' +
    '----{shaderTags} ------\n {shaderDescription}';
```

then when building readme contents:

```
var readmeContent = '';
...

readmeContent = README_CONTENT_TEMPLATE
    .replace('{shaderName}', getShaderName())
    .replace('{shaderTags}', getShaderTags())
    .replace.....
```

please look [here](https://github.com/patuwwy/ShaderToy-Chrome-Plugin/blob/455dd154ba1855b1e3436c00e4a5110d3c877882/shadertoy-plugin-profile.js#L32)

Reply:
Thanks for the link - I hadn't seen that.  Since we will have common code, OK if I abstract out a template-applying function?  Like `render_template(README_TEMPLATE,{shaderName: getShaderName(), shaderTags: getShaderTags()})` or some such?